sound/soc/sof/topology.c

```diff
@@ -1382,6 +1382,11 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 
 	w->dobj.private = swidget;
 	list_add(&swidget->list, &sdev->widget_list);
+
+	/* load module library if needed */
+	if (sdev->ipc->ops->fw_loader->load_library)
+		return sdev->ipc->ops->fw_loader->load_library(sdev, swidget);
```

Comment:
I still find it confusing that we have both

sof_ops(sdev)->load_library
sdev->ipc->ops->fw_loader->load_library(sdev, swidget);

Given 2 choices, half of engineers will get it 'wrong'

Reply:
And the error handling is incorrect as well, this section should be like this:
```c
	/* load module library if needed */
	if (sdev->ipc->ops->fw_loader->load_library) {
		ret = sdev->ipc->ops->fw_loader->load_library(sdev, swidget);
		if (ret) {
			dev_err(scomp->dev, "library loading for widget %s failed\n",
				swidget->widget->name);
			kfree(swidget->private);
			kfree(swidget->tuples);
			kfree(swidget);
			return ret;
		}
	}

	w->dobj.private = swidget;
	list_add(&swidget->list, &sdev->widget_list);
```